### ISSUE TITLE
Is it possible to modify an argument to a USDT tracepoint

### ISSUE THREAD

**Issue by ghaidinyak** [iovisor/bcc#1353]
Is there a mechanism that a eBPF program can communicate with the USDT that it's attached to? Is it possible to modify an argument to the USDT?

**Comment by palmtenor** [comment]
There is the `bpf_probe_write_user` helper. But in order to use it for USDT we'll need to have generated code for that when parsing USDT args.

**Comment by yonghong-song** [comment]
Actually, the answer is no if you want to modify an USDT argument. The argument (the context) of a tracepoint bpf program (the structure members as defined in tracepoint format) is readonly. What is your use case here?

**Comment by ghaidinyak** [comment]
The use case for this request is to have an a priori agreed upon structure (a context). Then the eBPF program would examine the context compared to it's "comparison" context. If the context matches the comparison, then this match status would be passed back to the user program. When the user program receives a "match" status, then the user program can take addition actions.

Using this a different dynamic control system can be implemented, some thing that is unique to that trace point at that point in time. One use case would be LTTNG filters implemented with eBPF

**Comment by goldshtn** [comment]
@yonghong-song Our USDT implementation is based on uprobes. The arguments are not passed in a readonly structure like a tracepoint.

**Comment by yonghong-song** [comment]
@goldshtn Thanks for pointing out my mistake! Just checked the arguments (pt_regs) of uprobe program (kprobe type) are readonly as well. Just as an example,
(linux:samples/bpf/test_probe_write_user_kern.c, attaching to kprobe, but uprobe applies too)
...
int bpf_prog1(struct pt_regs *ctx) {
...
        struct sockaddr_in new_addr, orig_addr = {};
        void *sockaddr_arg = (void *)PT_REGS_PARM2(ctx); 
        ...
        bpf_probe_write_user(sockaddr_arg, &new_addr,
                                     sizeof(new_addr));
        ...
 }

Basically, you cannot modify argument itself. But if the argument is a pointer, you
change its contents.

**Comment by yonghong-song** [comment]
@ghaidinyak So if your context is pointed to by an argument pointer, yes, you can modify its contents.

**Comment by palmtenor** [comment]
@yonghong-song What I meant (and @goldshtn if I understand correctly) is that the USDT arguments are just user-space virtual addresses. Remember how we generate the `_bpf_readarg_PROBE_1` helpers? It's different from call-stack parameter (i.e. `PT_REGS_PARMX` from `ctx`)

But actually you are partially right. USDT arguments **might** also be just register values, which I'm pretty sure is readonly.

**Comment by ghaidinyak** [comment]
Thanks!

**Comment by goldshtn** [comment]
I believe this is addressed; if you have any further questions, please don't hesitate to task.
